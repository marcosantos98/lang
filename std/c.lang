extern "c" puts :: int(str char*)
extern "c" putchar :: int(d int)
extern "c" exit :: int(code int)
extern "c" printf :: int(str char*, args ...)
