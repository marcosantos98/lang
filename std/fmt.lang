import "std:c"

putd :: void(d int) {
    if d > 9 {
	var a = d / 10
	d -= 10 * a
	putd(a)
    }
    putchar(48 + d)
}
