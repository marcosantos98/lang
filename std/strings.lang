import "std:c"

struct StringBuilder {
    data ^char,
    len int,
    cap int,
    a string
}

sb_new :: StringBuilder() {
    var data = malloc(1024)
    memset(data, 0, 1024)
    return StringBuilder{data as ^char, 0, 1024}
}

sb_write_str :: void(sb ^StringBuilder, append string) {
    var new_len = sb.len + len(append)
    if new_len > sb.cap {
	sb.cap = sb.cap * 2
	sb.data = realloc(sb.data, sb.cap) as ^char
    }
    var ptr: ^char = sb.data as ^char
    memcpy(ptr + sb.len, append.data, len(append))
    sb.len = new_len
}

sb_print :: void(sb StringBuilder) {
    printf(c("%.*s"), sb.len, sb.data)
}

sb_to_cstr :: ^char(sb StringBuilder) {
    return sb.data as ^char
}

sb_destroy :: void(sb StringBuilder) {
    free(sb.data)
}
