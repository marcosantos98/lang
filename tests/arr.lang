import "std:c"

add :: void(a int*) {
    *a += 1
}

print :: void(a int) {
    printf(xx to_cstring("%d\n"), a)
}

main :: void() {
    var ns = [10]int{0, 2}
    ns[0] = 0
    printf(xx to_cstring("%d\n"), ns[0])
    add(&ns[0])
    printf(xx to_cstring("%d\n"), ns[0])
    var z = &ns[1]
    print(*z)
    *z += 1
    printf(xx to_cstring("%d\n"), *z)
}
