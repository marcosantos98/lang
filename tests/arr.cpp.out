/* -- Builtin types -- */
typedef const char* cstr;
/* -- ------------- -- */
/* --------------- Builtin ---------------- */
struct String {
    void* data;
    int len;
};
static String ___empty_string___ = {(void*)"", 0};

String make_string(const char* cstr, int len) {
    return {(void*)cstr, len};
}

int len_str(String s) {
    return s.len;
}
/* --------------- Builtin ---------------- */

extern "C" int puts(cstr);
extern "C" int putchar(int);
extern "C" int exit(int);
void putd(int);
void add(int*);
void print(int);
void ___entry___();

void putd(int d){
if (d>9){
int a = d/10;
d -= 10*a;
putd(a);
}
putchar(48+d);
}
void add(int* a){
*a += 1;
}
void print(int a){
putd(a);
}
void ___entry___(){
int ns[10] = {0,
2};
ns[0] = 0;
putd(ns[0]);
add(&ns[0]);
puts((cstr)___empty_string___.data);
putd(ns[0]);
int* z = &ns[1];
puts((cstr)___empty_string___.data);
print(*z);
*z += 1;
puts((cstr)___empty_string___.data);
putd(*z);
}
int main(int argc, char** argv) {
___entry___();
}

